mx.sulong/mx_sulong_benchmarks.py

```diff
@@ -0,0 +1,257 @@
+
+def _failurePatterns():
+    return [
+        re.compile(r'error:'),
+        re.compile(r'Exception')
+    ]
+
+def _successPatterns():
+    return [re.compile(r'^(### )?([a-zA-Z0-9\.\-_]+): +([0-9]+(?:\.[0-9]+)?)', re.MULTILINE)]
+
+
+
+class NativeBenchmarkSuite(StdOutBenchmarkSuite):
+    def group(self):
+        return _group()
+
+    def subgroup(self):
+        return _subgroup()
+
+    def benchmarkList(self, bmSuiteArgs):
+        return _benchmarkList(bmSuiteArgs)
+
+    def benchHigherScoreRegex(self):
+        return _benchHigherScoreRegex()
+
+    def failurePatterns(self):
+        return _failurePatterns()
+
+    def successPatterns(self):
+        return _successPatterns()
+
+    def vmArgs(self, bmSuiteArgs):
+        return self.parseCommandLineArgs(bmSuiteArgs)[0]
+
+    def runArgs(self, bmSuiteArgs):
+        return self.parseCommandLineArgs(bmSuiteArgs)[1]
+
+    def parseCommandLineArgs(self, bmSuiteArgs):
+        vmArgs, runArgs = mx_benchmark.splitArgs(bmSuiteArgs, "--")
+
+        mx.logv('vmArgs: %s' % str(vmArgs))
+        mx.logv('runArgs: %s' % str(runArgs))
+
+        return vmArgs, runArgs
+
+    def before(self, bmSuiteArgs):
+        self.currentDir = os.getcwd()
+        os.chdir(_benchmarksDirectory())
+
+    def after(self, bmSuiteArgs):
+        os.chdir(self.currentDir)
+
+    def rules(self, out, benchmarks, bmSuiteArgs):
+        return [
+          mx_benchmark.StdOutRule(_benchHigherScoreRegex(), {
+                "vm": "sulong",
+                "benchmark": ("<benchmark>", str),
+                "metric.name": "score",
+                "metric.type": "numeric",
+                "metric.value": ("<score>", float),
+                "metric.score-function": "id",
+                "metric.better": "higher",
+                "metric.iteration": 0,
+                "host-vm": "native",
+                "host-vm-config": "none",
+                "guest-vm": self.getGuestVMName(),
+                "guest-vm-config": self.getGuestVMConfig(),
+          }),
+        ]
+
+class GCCBenchmarkSuite(NativeBenchmarkSuite):
+
+    def name(self):
+        return 'gcc'
+
+    def getGuestVMName(self):
+        return 'gcc'
+
+    def getGuestVMConfig(self):
+        return 'O3'
+
+    def runAndReturnStdOut(self, benchmarks, bmSuiteArgs):
+        inputFile = benchmarks[0]
+        _, ext = os.path.splitext(benchmarks[0])
+        f = open(os.devnull, 'w')
+        if ext == '.c':
+            mx.run(['gcc', '-std=gnu99'] + bmSuiteArgs + ['-lm', '-lgmp', '-O3'] + [inputFile], out=f, err=f)
+        elif ext == '.cpp':
+            mx.run(['g++'] + bmSuiteArgs + ['-lm', '-lgmp', '-O3'] + [inputFile], out=f, err=f)
+        else:
+            exit(ext + " is not supported!")
+
+        myStdOut = mx.OutputCapture()
+        retCode = mx.run(['./a.out'], out=myStdOut, err=f)
+        return [retCode, myStdOut.data]
+
+class CLANGBenchmarkSuite(NativeBenchmarkSuite):
```

Comment:
The idea was that there is one `BenchmarkSuite` per benchmark suite, this `BenchmarkSuite` then should delegate to other objects for execution.
To take a java-world example, one "DaCapo" `BenchmarkSuite` class that delegates to a `JavaVM` class for execution. This benchmark suite then has options (`--jvm` and `--jvm-config`) that allow it to pick the right `JavaVM`. Then the benchmark suite composes the dacapo-specific part for the command line (e.g., `["-jar", "dacapo.jar", "fop"]`) and asks the `JavaVM` to launch a java VM with those args. `JavaVM` can build the rest of the command line to get "HotSpot configured with Graal", or "HotSpot configured with C2" etc.

So you could have such as `NativeVM` class with implementations for GCC, clang and Sulong.

In the case of hosted VMs such as Sulong you need an additional layer of delegation since such a `SulongVM` would have to delegate to a `JavaVM` for the Java part.